File: archive/c/c/roman-numeral.c
#include  <stdio.h>
#include <stdlib.h>
#include <string.h>

int val[150];

int main(int argc,char **argv)
{
    if(argv[1]==NULL){
        printf("Usage: please provide a string of roman numerals");
        return 0;
    }
    if(strlen(argv[1])==0){
        printf("0");
        return 0;
    }
    
    int len=strlen(argv[1]);
    val['I']=1;
    val['V']=5;
    val['X']=10;
    val['L']=50;
    val['C']=100;
    val['D']=500;
    val['M']=1000;
    long long ans=0;
    
    for(int i=1;i<len;++i){
        if(!val[argv[1][i]]){
            printf("Error: invalid string of roman numerals");
            return 0;
        }
        if(val[argv[1][i]]>val[argv[1][i-1]])ans-=2*val[argv[1][i-1]];
        ans+=val[argv[1][i]];
    }
    if(!val[argv[1][0]]){
        printf("Error: invalid string of roman numerals");
        return 0;
    }
    printf("%lld",ans+val[argv[1][0]]);
}

File: archive/r/ruby/roman-numeral.rb
ROMAN = {
  "I" => 1, "V" => 5, "X" => 10, "L" => 50,
  "C" => 100, "D" => 500, "M" => 1000
}.freeze

USAGE = "Usage: please provide a string of roman numerals"
ERROR = "Error: invalid string of roman numerals"

def valid_roman?(s)
  return false unless s.match?(/\A[IVXLCDM]+\z/)

  # no more than 3 repeats
  return false if s.match?(/(.)\1{3,}/)

  # invalid repeatables
  %w[V L D].each { |ch| return false if s.include?(ch * 2) }

  true
end

def roman_to_int(s)
  return USAGE if s.nil?

  s = s.upcase
  return 0 if s.empty?
  return ERROR unless valid_roman?(s)

  total = 0

  s.chars.each_cons(2) do |a, b|
    total += (ROMAN[a] < ROMAN[b]) ? -ROMAN[a] : ROMAN[a]
  end

  total + ROMAN[s[-1]]
end

puts roman_to_int(ARGV.first)
